nightly-2024-04-05 regression: error[E0275]: overflow evaluating the requirement

### Code

Sorry, I don't have simple code to reproduce the issue, because the project is large.

The repro is this:

```
$ git clone https://github.com/facebook/starlark-rust

$ git co 96a2bf3d164d99cccc6cf330433d770e7d01b15a  # Latest revision

$ cargo +nightly-2024-04-05-aarch64-apple-darwin test -p starlark --lib --no-run

This works (warnings are unrelated)

$ cargo +nightly-2024-04-06-aarch64-apple-darwin test -p starlark --lib --no-run

...

error[E0275]: overflow evaluating the requirement `&_: StarlarkTypeRepr`
    |
    = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`starlark`)
note: required for `&[_]` to implement `StarlarkTypeRepr`
   --> starlark/src/values/types/list/value.rs:251:17
    |
251 | impl<'a, V: 'a> StarlarkTypeRepr for &'a [V]
    |                 ^^^^^^^^^^^^^^^^     ^^^^^^^
252 | where
253 |     &'a V: StarlarkTypeRepr,
    |            ---------------- unsatisfied trait bound introduced here
    = note: 126 redundant requirements hidden
    = note: required for `&[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[...]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]` to implement `StarlarkTypeRepr`
    = note: the full name for the type has been written to '/Users/nga/devel/starlark-rust/target/debug/deps/starlark-e78de373a7652cbb.long-type-11371795697589288004.txt'
```

Contents of the mentioned file is:

```
&[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[_]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
```

I tried increasing `#![recursion_limit = "256"]`, did not help.

### Version it worked on

```
$ cargo +nightly-2024-04-05-aarch64-apple-darwin --version
cargo 1.79.0-nightly (0637083df 2024-04-02)
$ rustc +nightly-2024-04-05-aarch64-apple-darwin --version
rustc 1.79.0-nightly (385fa9d84 2024-04-04)
```

Also works on stable.

### Version with regression

```
$ cargo +nightly-2024-04-06-aarch64-apple-darwin --version
cargo 1.79.0-nightly (0637083df 2024-04-02)
$ rustc +nightly-2024-04-06-aarch64-apple-darwin --version
rustc 1.79.0-nightly (9d79cd5f7 2024-04-05)
```

This may be duplicate of https://github.com/rust-lang/rust/issues/123493 because timing.

This is also affecting `diesel`:

```rust
error[E0275]: overflow evaluating the requirement `(&_, &_, &_, ..., ..., ...): diesel::Insertable<...>`
  |
  = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`rust_out`)
  = note: required for `&(_, _, _, _, _, _)` to implement `diesel::Insertable<table>`
  = note: 127 redundant requirements hidden
  = note: required for `(&((..., ..., ..., ..., ..., ...), ..., ..., ..., ..., ...), ..., ..., ..., ..., ...)` to implement `diesel::Insertable<table>`
  = note: the full name for the type has been written to '/tmp/rustdoctestmnF1pz/rust_out.long-type-2635572469679000095.txt'
  = note: consider using `--verbose` to print the full type name to the console

error: aborting due to 1 previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0275`.
Couldn't compile the test.

failures:
    contrib/db_pools/lib/src/diesel.rs - diesel (line 26)
```

--- a/compiler/rustc_lint/src/non_local_def.rs
+++ b/compiler/rustc_lint/src/non_local_def.rs
@@ -265,7 +265,19 @@ fn impl_trait_ref_has_enough_non_local_candidates<'tcx>(
     binder: EarlyBinder<TraitRef<'tcx>>,
     mut did_has_local_parent: impl FnMut(DefId) -> bool,
 ) -> bool {
-    let infcx = tcx.infer_ctxt().build();
+    let infcx = tcx
+        .infer_ctxt()
+        // We use the new trait solver since the obligation we are trying to
+        // prove here may overflow and those are fatal in the old trait solver.
+        // Which is unacceptable for a lint.
+        //
+        // Thanksfully the part we use here are very similar to the
+        // new-trait-solver-as-coherence, which is in stabilization.
+        //
+        // https://github.com/rust-lang/rust/issues/123573
+        .with_next_trait_solver(true)
+        .build();
+
     let trait_ref = binder.instantiate(tcx, infcx.fresh_args_for_item(infer_span, trait_def_id));
 
     let trait_ref = trait_ref.fold_with(&mut ReplaceLocalTypesWithInfer {

--- a/tests/ui/lint/non-local-defs/trait-solver-overflow-123573.rs
+++ b/tests/ui/lint/non-local-defs/trait-solver-overflow-123573.rs
@@ -1,0 +1,14 @@
+//@ check-pass
+//@ edition:2021
+
+// https://github.com/rust-lang/rust/issues/123573#issue-2229428739
+
+pub trait Test {}
+
+impl<'a, T: 'a> Test for &[T] where &'a T: Test {}
+
+fn main() {
+    struct Local {}
+    impl Test for &Local {}
+    //~^ WARN non-local `impl` definition
+}

--- a/tests/ui/lint/non-local-defs/trait-solver-overflow-123573.stderr
+++ b/tests/ui/lint/non-local-defs/trait-solver-overflow-123573.stderr
@@ -1,0 +1,14 @@
+warning: non-local `impl` definition, they should be avoided as they go against expectation
+  --> $DIR/trait-solver-overflow-123573.rs:12:5
+   |
+LL |     impl Test for &Local {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = help: move this `impl` block outside the of the current function `main`
+   = note: an `impl` definition is non-local if it is nested inside an item and may impact type checking outside of that item. This can be the case if neither the trait or the self type are at the same nesting level as the `impl`
+   = note: one exception to the rule are anon-const (`const _: () = { ... }`) at top-level module and anon-const at the same nesting as the trait or type
+   = note: this lint may become deny-by-default in the edition 2024 and higher, see the tracking issue <https://github.com/rust-lang/rust/issues/120363>
+   = note: `#[warn(non_local_definitions)]` on by default
+
+warning: 1 warning emitted
+

